### ISSUE TITLE
Circuit Breaker: Use a cached value during the open state instead of rethrowing exceptions

### ISSUE THREAD

**Issue by pampua84** [App-vNext/Polly#1089]
Hi, as the title suggests, it would be interesting to use a cached value during the Circuit Breaker's open state instead of having it re-raise the exceptions, because even if, as written in the documentation, "failing fast is better than making users/callers wait." even exceptions have their cost and if the number of these rises significantly, they could still lead to an overload of the host where the application is present.
Is it possible to achieve this result? Is it a feature that could arouse interest?
Thank you

**Comment by PeterCsalaHbo** [comment]
TL;DR: Circuit Breaker's transient error detection scope is a service, not an endpoint. Caching should be applied case-by-case where you put into the account many factors.
___
### Circuit Breaker
Circuit Breaker is a resiliency primitive which can be used to detect that a given downstream system is malfunctioning (reactive) and it protects the upstream systems from transient errors (proactive).

In distributed systems it is pretty hard to be sure that a given service is down. It could be just a temporal network partitioning problem. It could be simply just overload of the system.  Or any other issue. From a client perceptive the observable behaviour is that some of the outgoing requests are failing. 

Circuit Breaker as a proxy can measure the success - failure rate and proactively can prohibit new requests to give space and time for the downstream system for self-healing. Since the Circuit Breaker is a shared resource many different requests against the same downstream can go trough it.

My point here is that you are using CB against downstream service, not on endpoint level. In extreme cases where we are talking about nano services (single exposed endpoint) a CB can detect failure on endpoint basis. But that's not common.

### Cache
Caching is a hard problem. What should be cached? Is the data cacheable at all? Until when should we store the cached data? When and how we update cache data? What if the cache is full? What if there is a cache miss? etc...

My point here is that you should carefully evaluate each and every case one-by-one. As it said in the previous section, CB is a shared resource and many different requests can go through it. Does it make sense to cache POST request's response? Does it make sense to cache all GET request's response? Should we delete all cached value if the CB transitions to closed state again? Should we still throw an exception is there is a cache miss?

We can't define a behaviour which is suitable for all situation. 

### Combining the two 
Polly offers a set of resiliency primitives which can be used individually or they can be combined to define a resiliency strategy. Polly provides a built-in solution for policy chaining (`PolicyWrap`) or by there are [many other ways](https://stackoverflow.com/questions/42093593/execute-multiple-policies/73106630#73106630) as well.

In this particular case you can define a cache and a cb policies separately. And you can use the [CircuitState](https://github.com/App-vNext/Polly/wiki/Circuit-Breaker#circuit-state-for-health-reporting-polly-v40) to branch your logic. If it is open then you can try to retrieve cached data otherwise let the request go through the CB proxy.

**Comment by pampua84** [comment]
Hi Peter, thanks to your advice and I managed to implement a similar solution that is right for me. Thank you
